File: app/vectorbt_test/core/signals.py
import enum
from abc import ABC
from vectorbt_test.core.nodes import FeatureNode, NodeType, NodeDType, to_node
from vectorbt_test.core.node_builder import NodeBuilder
from vectorbt_test.core.registry import NodeRegistry, NodeMeta, NodeParam
from vectorbt_test.core.context import PortfolioContext
import pandas as pd


class SignalScope(enum.Enum):
    Null = -1
    TS = 1
    CS = 2
    TS_CS = 4


class Signal(FeatureNode, ABC):
    @staticmethod
    def build(expr_str: str):
        node = NodeBuilder().build(expr_str)
        assert node.is_signal
        return node

    def __init__(self, *args):
        super().__init__(NodeType.Signal, NodeDType.Signal)
        self.scope = SignalScope.Null
        self.args = [to_node(a) for a in args]

    def _args(self):
        return [a.cache_key() for a in self.args]
  
    def is_scope(self, scope: int) -> bool:
        return (self.scope.value & scope) == self.scope.value
    
    def __and__(self, other):
        return BinarySignalOp(self, other, op="and")

    def __or__(self, other):
        return BinarySignalOp(self, other, op="or")

    def when(self, schedule):
        return SignalGate(self, schedule)

    def filter(self, condition):
        return SignalGate(self, condition)

    def confirm(self, condition):
        return SignalGate(self, condition)

    def gate(self, other):
        return SignalGate(self, other)

    def cross(self, b):
        return Cross(self, b)
    
    def crossunder(self, b):
        return CrossUnder(self, b)
    
    def cooldown(self, n):
        return Cooldown(self, n)

    def hold(self, n):
        return Hold(self, n)


class SignalGate(Signal):
    def __init__(self, signal, signal_gate):
        super().__init__()
        self.scope = SignalScope.TS_CS
        self.signal = signal
        self.signal_gate = signal_gate

    def compute(self, data, context: PortfolioContext):
        s = self.signal.evaluate(data, context)
        g = self.signal_gate.evaluate(data, context)

        return s & g


class TSSignal(Signal):
    def __init__(self):
        super().__init__()
        self.scope = SignalScope.TS


class Cooldown(TSSignal):
    def __init__(self, signal, n):
        super().__init__()
        self.signal = signal
        self.n = n

    def compute(self, data, context: PortfolioContext):
        s = self.signal.evaluate(data, context).fillna(False)

        result = s.copy()
        cooldown = 0

        for i in range(len(s)):
            if cooldown > 0:
                result.iloc[i] = False
                cooldown -= 1
            elif s.iloc[i]:
                cooldown = self.n - 1  # 当前算1天
                result.iloc[i] = True

        return result


class Hold(TSSignal):
    def __init__(self, signal, n):
        super().__init__()
        self.signal = signal
        self.n = n

    def compute(self, data, context: PortfolioContext):
        s = self.signal.evaluate(data, context).fillna(False)

        result = pd.Series(False, index=s.index)

        hold = 0

        for i in range(len(s)):
            if s.iloc[i]:
                hold = self.n
            if hold > 0:
                result.iloc[i] = True
                hold -= 1

        return result
    

class BinarySignalOp(Signal):
    def __init__(self, left, right, op):
        super().__init__()
        self.scope = SignalScope.TS_CS
        self.left = left
        self.right = right
        self.op = op

    def is_scope(self, scope: int) -> bool:
        return (super().is_scope(scope)
                and self.left.is_scope(scope)
                and self.right.is_scope(scope))
    
    def compute(self, data, context: PortfolioContext):
        l = self.left.evaluate(data, context)
        r = self.right.evaluate(data, context)

        if self.op == "and":
            return l & r
        elif self.op == "or":
            return l | r


class Cross(Signal):
    def __init__(self, left, right):
        super().__init__()
        self.scope = SignalScope.TS_CS
        self.left = left
        self.right = right

    def compute(self, data, context: PortfolioContext):
        a = self.left.evaluate(data, context)
        b = self.right.evaluate(data, context)

        # return (a > b) & (a.shift(1) <= b.shift(1))
        return self.apply(
            a,
            lambda x: (x > b.loc[x.index]) & (x.shift(1) <= b.loc[x.index].shift(1)),
            context
        )


class CrossUnder(Signal):
    def __init__(self, left, right):
        super().__init__()
        self.scope = SignalScope.TS_CS
        self.left = left
        self.right = right

    def compute(self, data, context: PortfolioContext):
        a = self.left.evaluate(data, context)
        b = self.right.evaluate(data, context)

        return (a < b) & (a.shift(1) >= b.shift(1))


class CSSignal(Signal):
    def __init__(self):
        super().__init__()
        self.scope = SignalScope.CS


class RebalanceDaily(CSSignal):
    def compute(self, data, context: PortfolioContext):
        return pd.Series(True, index=data.index)


class RebalanceWeekly(CSSignal):

    def __init__(self, weekday=0):  # 0=Monday
        super().__init__()
        self.weekday = weekday

    def compute(self, data, context: PortfolioContext):
        dt = data.index
        return pd.Series(dt.weekday == self.weekday, index=dt)
    

class RebalanceMonthly(CSSignal):
    def __init__(self, day=1):
        super().__init__()
        self.day = day

    def compute(self, data, context: PortfolioContext):
        dt = data.index
        return pd.Series(dt.day == self.day, index=dt)


class RebalanceEveryNDays(CSSignal):
    def __init__(self, n):
        super().__init__()
        self.n = n

    def compute(self, data, context: PortfolioContext):
        idx = data.index
        mask = pd.Series(False, index=idx)
        mask.iloc[::self.n] = True
        return mask


class RebalanceOnDates(CSSignal):
    def __init__(self, dates):
        super().__init__()
        self.dates = pd.to_datetime(dates)

    def compute(self, data, context: PortfolioContext):
        idx = data.index
        return pd.Series(idx.isin(self.dates), index=idx)


class RebalanceMonthEnd(CSSignal):
    def compute(self, data, context: PortfolioContext):
        idx = data.index
        df = pd.Series(index=idx, data=False)

        month = idx.to_period("M")
        last_days = idx.to_series().groupby(month).idxmax()

        df.loc[last_days.values] = True
        return df  


class RebalanceWeekEnd(CSSignal):
    def compute(self, data, context: PortfolioContext):
        idx = data.index
        df = pd.Series(False, index=idx)

        week = idx.to_period("W")
        last_days = idx.to_series().groupby(week).idxmax()

        df.loc[last_days.values] = True
        return df


NodeRegistry.register(
    "Cooldown",
    lambda signal, n: Cross(signal, n),
    NodeMeta(
        name="Cooldown",
        group="signal",
        desc="Cooldown Signal",
        params=[
            NodeParam("signal", "Signal", desc="Signal"),
            NodeParam("n", "int", desc="number")
        ]
    ))


NodeRegistry.register(
    "Hold",
    lambda signal, n: Hold(signal, n),
    NodeMeta(
        name="Hold",
        group="signal",
        desc="Hold Signal",
        params=[
            NodeParam("signal", "Signal", desc="Signal"),
            NodeParam("n", "int", desc="hold days")
        ]
    ))

NodeRegistry.register(
    "Cross",
    lambda left, right: Cross(left, right),
    NodeMeta(
        name="Cross",
        group="signal",
        desc="Corss Signal",
        params=[
            NodeParam("left", "Node", desc="左节点"),
            NodeParam("right", "Node", desc="右节点")
        ]
    ))

NodeRegistry.register(
    "CrossUnder",
    lambda left, right: CrossUnder(left, right),
    NodeMeta(
        name="CrossUnder",
        group="signal",
        desc="CrossUnder Signal",
        params=[
            NodeParam("left", "Node", desc="左节点"),
            NodeParam("right", "Node", desc="右节点")
        ]
    ))

NodeRegistry.register(
    "RebalanceDaily",
    lambda: RebalanceDaily(),
    NodeMeta(
        name="RebalanceDaily",
        group="signal",
        desc="RebalanceDaily Signal"
    ))

NodeRegistry.register(
    "RebalanceWeekly",
    lambda weekday=0: RebalanceWeekly(weekday),
    NodeMeta(
        name="RebalanceWeekly",
        group="signal",
        desc="RebalanceWeekly Signal",
        params=[
            NodeParam("weekday", "int", 0, "weekday")
        ]
    ))

NodeRegistry.register(
    "RebalanceMonthly",
    lambda day=1: RebalanceMonthly(day),
    NodeMeta(
        name="RebalanceMonthly",
        group="signal",
        desc="RebalanceMonthly Signal",
        params=[
            NodeParam("day", "int", 1, "day")
        ]   
    ))

NodeRegistry.register(
    "RebalanceEveryNDays",
    lambda n: RebalanceEveryNDays(n),
    NodeMeta(
        name="RebalanceEveryNDays",
        group="signal",
        desc="RebalanceEveryNDays Signal",
        params=[
            NodeParam("n", "int", 1, "n")
        ]   
    ))

NodeRegistry.register(
    "RebalanceOnDates",
    lambda dates: RebalanceOnDates(dates),
    NodeMeta(
        name="RebalanceOnDates",
        group="signal",
        desc="RebalanceOnDates Signal",
        params=[
            NodeParam("dates", "list", [], "dates")
        ]   
    ))

NodeRegistry.register(
    "RebalanceMonthEnd",
    lambda: RebalanceMonthEnd(),
    NodeMeta(
        name="RebalanceMonthEnd",
        group="signal",
        desc="RebalanceMonthEnd Signal"
    ))

NodeRegistry.register(
    "RebalanceWeekEnd",
    lambda: RebalanceWeekEnd(),
    NodeMeta(
        name="RebalanceWeekEnd",
        group="signal",
        desc="RebalanceWeekEnd Signal"
    ))

# NodeRegistry.register("cross", lambda: CrossSignalNode(), group="signal")
# NodeRegistry.register("breakout", lambda: BreakoutSignalNode(), group="signal")

File: quant-mantis/src/model/dsl_node/signals.ts
import { NodeType, NodeDType, PortfolioContext } from "./types";
import { Node, FeatureNode } from "./nodes";

export enum SignalGroup {
    Null = -1,
    TS = 1,
    CS = 2,
    TS_CS = 4,
}

// 信号组枚举已在 types.ts 定义
export abstract class Signal extends FeatureNode {
    signalGroup: SignalGroup;

    constructor(name: string | null = null, signalGroup: SignalGroup = SignalGroup.Null) {
        super(NodeType.Signal);
        this.dtype = NodeDType.SIGNAL;
        this._name = name || this.constructor.name;
        this.signalGroup = signalGroup;
    }

    static build(_nodeExpr: string | any | Node): Signal {
        // 忽略构建逻辑，返回空实现
        return new SignalWrapper(null, {} as Node);
    }

    get name(): string {
        return this._name;
    }

    _args(): any[] {
        return [this.signalGroup, ...super._args()];
    }

    isGroup(signalGroups: number): boolean {
        return (this.signalGroup & signalGroups) === this.signalGroup;
    }

    and(other: Signal): BinarySignalOp {
        return new BinarySignalOp(this, other, "and");
    }

    or(other: Signal): BinarySignalOp {
        return new BinarySignalOp(this, other, "or");
    }

    when(schedule: Signal): SignalGate {
        return new SignalGate(this, schedule);
    }

    filter(condition: Signal): SignalGate {
        return new SignalGate(this, condition);
    }

    confirm(condition: Signal): SignalGate {
        return new SignalGate(this, condition);
    }

    gate(other: Signal): SignalGate {
        return new SignalGate(this, other);
    }

    cross(b: FeatureNode): Cross {
        return new Cross(this, b);
    }

    crossunder(b: FeatureNode): CrossUnder {
        return new CrossUnder(this, b);
    }

    cooldown(n: number): Cooldown {
        return new Cooldown(this, n);
    }

    hold(n: number): Hold {
        return new Hold(this, n);
    }
}

// 信号包装器
export class SignalWrapper extends Signal {
    node: Node;

    constructor(name: string | null, node: Node, group: SignalGroup = SignalGroup.TS_CS) {
        super(name, group);
        if (node.dtype !== NodeDType.BOOL) {
            throw new Error("wrong node dtype, expect bool type for Signal");
        }
        this.dtype = node.dtype;
        this.node = node;
    }

    _args(): any[] {
        return [this.node.cacheKey(), ...super._args()];
    }

    compute(_data: any, _context: PortfolioContext): any {
        return null;
    }
}

// 时间序列信号
export class TSSignal extends Signal {
    constructor() {
        super(null, SignalGroup.TS);
    }
}

// 冷却信号
export class Cooldown extends TSSignal {
    signal: Signal;
    n: number;

    constructor(signal: Signal, n: number) {
        super();
        this.signal = signal;
        this.n = n;
    }

    _args(): any[] {
        return [this.signal.cacheKey(), this.n, ...super._args()];
    }

    compute(_data: any, _context: PortfolioContext): any {
        return null;
    }
}

// 持有信号
export class Hold extends TSSignal {
    signal: Signal;
    n: number;

    constructor(signal: Signal, n: number) {
        super();
        this.signal = signal;
        this.n = n;
    }

    _args(): any[] {
        return [this.signal.cacheKey(), this.n, ...super._args()];
    }

    compute(_data: any, _context: PortfolioContext): any {
        return null;
    }
}

// 复合信号
export abstract class ComplexSignal extends Signal {
    left: Signal;
    right: Signal;

    constructor(left: Signal, right: Signal) {
        super();
        this.left = left;
        this.right = right;
    }

    _args(): any[] {
        return [this.left.cacheKey(), this.right.cacheKey(), ...super._args()];
    }
}

// 信号门控
export class SignalGate extends ComplexSignal {
    constructor(signal: Signal, signalGate: Signal) {
        super(signal, signalGate);
        this.signalGroup = SignalGroup.TS_CS;
    }

    compute(_data: any, _context: PortfolioContext): any {
        return null;
    }
}

// 二进制信号操作
export class BinarySignalOp extends ComplexSignal {
    op: string;

    constructor(left: Signal, right: Signal, op: string) {
        super(left, right);
        this.signalGroup = SignalGroup.TS_CS;
        this.op = op;
    }

    _args(): any[] {
        return [this.left.cacheKey(), this.right.cacheKey(), this.op, ...super._args()];
    }

    compute(_data: any, _context: PortfolioContext): any {
        return null;
    }
}

// 上穿信号
export class Cross extends ComplexSignal {
    constructor(left: FeatureNode, right: FeatureNode) {
        super(left as Signal, right as Signal);
        if (left.dtype !== NodeDType.NUMERIC || right.dtype !== NodeDType.NUMERIC) {
            throw new TypeError("Cross requires numeric inputs");
        }
        this.signalGroup = SignalGroup.TS_CS;
    }

    compute(_data: any, _context: PortfolioContext): any {
        return null;
    }
}

// 下穿信号
export class CrossUnder extends ComplexSignal {
    constructor(left: FeatureNode, right: FeatureNode) {
        super(left as Signal, right as Signal);
        this.signalGroup = SignalGroup.TS_CS;
    }

    compute(_data: any, _context: PortfolioContext): any {
        return null;
    }
}

// 截面信号
export class CSSignal extends Signal {
    constructor() {
        super(null, SignalGroup.CS);
    }
}

// 每日再平衡
export class RebalanceDaily extends CSSignal {
    compute(_data: any, _context: PortfolioContext): any {
        return null;
    }
}

// 每周再平衡
export class RebalanceWeekly extends CSSignal {
    weekday: number;

    constructor(weekday: number = 0) {
        super();
        this.weekday = weekday;
    }

    _args(): any[] {
        return [this.weekday, ...super._args()];
    }

    compute(_data: any, _context: PortfolioContext): any {
        return null;
    }
}

// 每月再平衡
export class RebalanceMonthly extends CSSignal {
    day: number;

    constructor(day: number = 1) {
        super();
        this.day = day;
    }

    _args(): any[] {
        return [this.day, ...super._args()];
    }

    compute(_data: any, _context: PortfolioContext): any {
        return null;
    }
}

// 每N天再平衡
export class RebalanceEveryNDays extends CSSignal {
    n: number;

    constructor(n: number) {
        super();
        this.n = n;
    }

    _args(): any[] {
        return [this.n, ...super._args()];
    }

    compute(_data: any, _context: PortfolioContext): any {
        return null;
    }
}

// 指定日期再平衡
export class RebalanceOnDates extends CSSignal {
    dates: Date[];

    constructor(dates: Date[]) {
        super();
        this.dates = dates;
    }

    _args(): any[] {
        return [this.dates, ...super._args()];
    }

    compute(_data: any, _context: PortfolioContext): any {
        return null;
    }
}

// 月末再平衡
export class RebalanceMonthEnd extends CSSignal {
    compute(_data: any, _context: PortfolioContext): any {
        return null;
    }
}

// 周末再平衡
export class RebalanceWeekEnd extends CSSignal {
    compute(_data: any, _context: PortfolioContext): any {
        return null;
    }
}

// 注册函数（简化）
export function registerSignals(): void {
    // 忽略注册逻辑
}